apps/predictions/lib/stream_topic.ex

```diff
@@ -0,0 +1,67 @@
+defmodule Predictions.StreamTopic do
+  @moduledoc """
+  Translation between `SiteWeb.PredictionsChannel` topics, `Predictions.Store`
+  and `Predictions.Stream`.
+
+  The topic name joined in `SiteWeb.PredictionsChannel` is the input which
+  determines
+  - the appropriate `fetch_keys` needed to get relevant predictions from the
+    `Predictions.Store`
+  - one or more `Predictions.Stream` instances needed to fetch the relevant
+    predictions.
+
+  Every topic will be matched with one or more V3 API calls via
+  `Predictions.Stream` which filter the `/predictions` streaming endpoint on a
+  route and direction.
+  """
+
+  alias Predictions.Store
+  alias Routes.Route
+  alias Stops.Stop
+
+  defstruct [:topic, :fetch_keys, :streams]
+
+  @type t :: %__MODULE__{
+          topic: String.t(),
+          fetch_keys: Store.fetch_keys(),
+          streams: [String.t()]
+        }
+
+  @type api_filter_t :: %{
+          required(:route) => Route.id_t(),
+          required(:direction) => 0 | 1
+        }
+
+  @spec new(String.t()) :: t() | {:error, term()}
+  def new(topic) do
+    with "stop:" <> stop_id when stop_id != "" <- topic,
+         filters when filters != [] <- filters(topic) do
```

Comment:
suggestion: you could pass `stop_id` in directly since you already matched it on line 37
```suggestion
         filters when filters != [] <- filters(stop_id) do
```
or alternatively, pull this into the function head:
```elixir
def new(topic)
def new("stop:" <> stop_id) when stop_id != "" do
  filters = stop_id |> RoutePatterns.Repo.by_stop_id() |> ...
  ...
end
def new(_) do
  {:error, :unsupported_topic}
end
```

Reply:
I am also trying to leave it structured so that we could expand it later (say, if we start to enable subscribing to prediction streams by a trip ID, or a route ID, etc.). I did end up adjusting this slightly.